cirq/google/calibration/workflow.py

```diff
@@ -64,20 +78,42 @@ class CircuitWithCalibration:
     moment_to_calibration: Sequence[Optional[int]]
 
 
-def prepare_floquet_characterization_for_moment(
+RequestT = TypeVar('RequestT', bound=PhasedFSimCalibrationRequest)
+
+
+def _deduce_calibration_request_type(
+    calibration_request_type: Optional[Type[RequestT]], options: PhasedFSimCalibrationOptions
```

Comment:
Yes, there's a variety of ways to organize this.

1. Have a mapping from Options to Request hidden in this `_deduce_calibration_request_type` helper
2. Have an attribute on `PhasedFSimCalibrationOptions` that is the calibration request type
3. Move all the `prepare_characterization_for_xxx` to abstract methods on the option
4. Have an abstract `create_request_for` which constructs the correct request

I think (1) and (3) are worse. (2) and (4) are better and pretty similar to each other. The method introduced with (4) would be identical to the constructor of the request type, so (2) would be more true to how the code works but maybe there's some additional logic you might want to put into the method for subclasses? Plus abstract methods might be more easily digestible than passing around class types as attributes.

Reply:
This was my first open question, which may not have been clear :)

Reply:
My preference would 4, 2, 1, 3.

(4) also clearly defines the responsibility and you can easily search for usage of this particular function. If you go with (2) this could be used in many crazy ways which might be problematic in the future.

Reply:
sounds good. changed